### PR TITLE
fix(controller): node message should be updated when lock msg changed

### LINKED ISSUE #13629: Templated mutex name not resolved
### Pre-requisites

- [X] I have double-checked my configuration
- [ ] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

The mutex name is not resolved correctly when it is based on a workflow parameter, which is also based on another workflow parameter.

See the workflow below.
1. mutex name is defined as `{{ workflow.parameters.derived-mutex-name }}`
2. `derived-mutex-name` is defined as `"{{ workflow.parameters.base-mutex-name }}"`
3. `base-mutex-name` is defined as a parameter of the argo submit command line, let's say `mycustommutex`

In this workflow dag is used (and might be the reason of the defect).

With this workflow, if I run the following command:
```
argo submit my-workflow.yml --parameter base-mutex-name=mycustommutex
```
 I get the following message indicating the name of the mutex at runtime:
```
Waiting for argo-wf/Mutex{{ workflow.parameters.base-mutex-name }} lock. Lock status: 0/1
```
However the command `echo {{ workflow.parameters.derived-mutex-name }}` outputs correctly the resolved name:
```
mycustommutex
```

I have not tested this scenario with the latest version, because I'm not responsible of the deployment of argo in my company.

### Version(s)

v3.5.8

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: mutex-name-
spec:
  entrypoint: main
  arguments:
    parameters:
    - name: base-mutex-name
    - name: derived-mutex-name
      value: "{{ workflow.parameters.base-mutex-name }}"
  templates:
  - name: main
    dag:
      tasks:
      - name: first
        template: print-mutex-name
      - name: second
        template: print-mutex-name
  - name: print-mutex-name
    synchronization:
      mutex:
        name: "{{ workflow.parameters.derived-mutex-name }}"
    container:
      image: alpine
      command: [ "/bin/sh", "-c" ]
      args:
      - |
        #!/bin/bash
        echo {{ workflow.parameters.derived-mutex-name }}
        sleep 30
```

### Logs from the workflow controller

```text
I'm not granted the permission to get these logs
```

### Logs from in your workflow's wait container

```text
I'm not granted the permission to get these logs
```

### PR DESCRIPTION
**Pull request by chengjoey**
Fixes #13629

### Motivation
When the sync lock is not released, the latest lock-msg is always used as the node message

### CODE REVIEW

**Comment by chengjoey** [comment]
/hold
add test

**Comment by MasonM** [comment]
@chengjoey FYI: I entered https://github.com/argoproj/argo-workflows/pull/13660 to fix the flaky tests. After that's merged, click the "Update branch" button at the bottom of the PR

**Comment by Joibel** [comment]
> /hold cancel

@chengjoey is this ready for review, I'm unsure of what you meant by this comment?

**Comment by chengjoey** [comment]
> > /hold cancel
> 
> @chengjoey is this ready for review, I'm unsure of what you meant by this comment?

yeah, it is ready for review, since I didn't add the test at first, I used `/hold cancel` to get ready after adding the test

File: workflow/controller/operator_test.go [hunk 1]
```diff
@@ -8512,13 +8517,18 @@ func TestMutexWfPendingWithNoPod(t *testing.T) {
 	ctx := context.Background()
 	controller.syncManager = sync.NewLockManager(GetSyncLimitFunc(ctx, controller.kubeclientset), func(key string) {
 	}, workflowExistenceFunc)
-	_, _, _, _, err := controller.syncManager.TryAcquire(wf, "test", &wfv1.Synchronization{Mutex: &wfv1.Mutex{Name: "welcome"}})
+	lockAcquired, _, msg, _, err := controller.syncManager.TryAcquire(wf, "test", &wfv1.Synchronization{Mutex: &wfv1.Mutex{Name: "welcome"}})
 	require.NoError(t, err)
 	woc := newWorkflowOperationCtx(wf, controller)
+	if !lockAcquired {
```
Comment (Joibel) [inline workflow/controller/operator_test.go:8523]: This feels wrong for a test.

Surely the outcome of `TryAquire` should always be the same for this test, and therefore we could assert that `!lockAcquired`? If lockAquired is true here we're not testing the new functionality.

**Review by Joibel — CHANGES_REQUESTED** [review]
As a follow up to this it would be possible to list all of the failing locks rather than just the first, so if you feel like doing that it would be useful.

File: workflow/controller/operator_test.go (hunk 1 above)
Comment (chengjoey) [reply workflow/controller/operator_test.go]: done~, I modified the test case and handed over the modification of msg and status to `operate`

File: workflow/controller/operator_test.go [hunk 2]
```diff
@@ -8510,15 +8512,20 @@ func TestMutexWfPendingWithNoPod(t *testing.T) {
 	cancel, controller := newController(wf)
 	defer cancel()
 	ctx := context.Background()
-	controller.syncManager = sync.NewLockManager(GetSyncLimitFunc(ctx, controller.kubeclientset), func(key string) {
```
Comment (Joibel) [inline workflow/controller/operator_test.go:8513]: Why has this been removed in this iteration?

File: workflow/controller/operator_test.go [hunk 3]
```diff
@@ -8510,15 +8512,20 @@ func TestMutexWfPendingWithNoPod(t *testing.T) {
...
+
+	// preempt lock
 	_, _, _, _, err := controller.syncManager.TryAcquire(wf, "test", &wfv1.Synchronization{Mutex: &wfv1.Mutex{Name: "welcome"}})
 	require.NoError(t, err)
 	woc := newWorkflowOperationCtx(wf, controller)
 
 	woc.operate(ctx)
 	assert.Equal(t, wfv1.WorkflowRunning, woc.wf.Status.Phase)
 	assert.Equal(t, wfv1.NodePending, woc.wf.Status.Nodes.FindByDisplayName("hello-world-mpdht").Phase)
+	assert.Equal(t, "Waiting for argo/Mutex/welcome lock. Lock status: 0/1", woc.wf.Status.Nodes.FindByDisplayName("hello-world-mpdht").Message)
```
Comment (Joibel) [inline workflow/controller/operator_test.go:8524]: :+1:

**Review by Joibel — CHANGES_REQUESTED** [review]

File: workflow/controller/operator_test.go (hunk 1 above)
Comment (Joibel) [reply workflow/controller/operator_test.go]: This part looks good.

File: workflow/controller/operator_test.go (hunk 2 above)
Comment (chengjoey) [reply workflow/controller/operator_test.go:8513]: syncManager will be created by `newController`, so maybe this is unnecessary

https://github.com/argoproj/argo-workflows/blob/5310c396ee1e9f4e5575f62ef5340a86540d8679/workflow/controller/controller_test.go#L331-L333

Comment (chengjoey) [reply workflow/controller/operator_test.go:8513]: thanks @Joibel , i got it. 
done

**Comment by Joibel** [comment]
This needs the linter issue fixing @chengjoey

**Comment by chengjoey** [comment]
> This needs the linter issue fixing @chengjoey

thanks @Joibel 

done

**Review by Joibel — APPROVED** [review]

**Pull request merged**